### DISCUSSION TITLE
Unable to get_to one particular class

### DISCUSSION THREAD

**Discussion by vulkan101** [nlohmann/json#2458] [General]
Note: solved! See Niels reply - I hadn't used a const json & in from_json.

I'm a bit stumped here. I am just getting started setting up json config export/import for my small test application. So far I have successfully exported and imported a couple of custom data types (my classes), I am aware of the namespace requirement, but I have one class that happens to be the main class for my application which for some reason the from_json is not being recognised.
E.g. my class is called TestApp::App.  A class called IO_json has a member _j which is a  nlohmann::json,
I can do this:
```
namespace TestApp {
  void IO_json::save(std::string filename, TestApp::App& app)
  {
    std::ofstream out(filename);
    
    _j["AppConfig"] = json(app);
    out << std::setw(4) << _j << std::endl;
  }
}
```
and to to_json is called correctly, output looks good. 
However I can not compile the following:
 ```
void IO_json::load(std::string filename, TestApp::App& app)
  {
    std::ifstream i(filename);
    json j = json::parse(i); 
    
    std::cout << j.dump(4) << '\n';
    
    auto a = j["AppConfig"];
    
    j.at("AppConfig").get_to(app);  // <- this won't compile, 

  }
```
I get the following:
```
error C2672: 'nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::get_to': no matching overloaded function found
1><SNIP)>TestApp\src\io_json.cpp(113,33): error C2783: 'ValueType &nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::get_to(ValueType &) noexcept(<expr>) const': **could not deduce template argument for '__formal'**
1><SNIP>ThirdParty\json\single_include\nlohmann\json.hpp(19589): message : see declaration of 'nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::get_to'
```

to and from json looks pretty standard:
  ```
namespace TestApp {
void to_json(json& j, const TestApp::App& app)
  {
    j = json{ {"camera", app.cam()} };
    auto meshes = app.meshes();    
    j["meshes"] = meshes;
  }
  void from_json(json& j, TestApp::App& app)
  {
    j.at("camera").get_to(app.cam());
  }
}
```

I have quite a few members in my TestApp class; I presumed that I don't need to define conversions for them all just yet? Any ideas where I can look to figure this out? 
I'm using the latest develop build and visual studio 2019 Version 16.7.6

**Comment by gregmarr** [comment]
`get_to` uses `from_json` not `to_json`.

**Reply by vulkan101** [reply]
did I say that it used to_json? I can't have been clear. The problem is with my trying to compile get_to. 
I can use to_json fine as below:
` _j["AppConfig"] = json(app);`
so that is *not* the problem.
When I try to use
`j.at("AppConfig").get_to(app);  // <- this won't compile, `
I can't even compile. so there may be an issue with this:
```
void from_json(json& j, TestApp::App& app)
{
  j.at("camera").get_to(app.cam());
}
```
but I can't see what it is.
I have implemented this succesfully for a bunch of other classes, but for some reason when trying to get_to for the app class there is an issue.

**Reply by nlohmann** [reply]
Can you try changing the signature to `void from_json(const json& j, TestApp::App& app)`

**Reply by vulkan101** [reply]
wow - facepalm - I can't believe I missed that... yep that fixes it! Nothing mysterious in the class at all, missed the const qualifier for the json object.

**Reply by nlohmann** [reply]
\o/

**Reply by gregmarr** [reply]
I only saw the `to_json`, guess I missed the `from_json` below it.  :)  Ah, I saw the `get_to` line and thought it was the line in main program that wasn't compiling.  Sorry for the confusion.
